pkg/scanfetcher/nexpose.go

```diff
@@ -0,0 +1,155 @@
+package scanfetcher
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"path"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/asecurityteam/nexpose-scan-notifier/pkg/domain"
+)
+
+const (
+	activeQueryParam = "active" // Bool indicating if active scans should be fetched.
+	pageQueryParam   = "page"   // The index of the page (zero-based) to retrieve.
+	sizeQueryParam   = "size"   // The number of records per page to retrieve.
+	sortQueryParam   = "sort"
+	sortQueryValue   = "endTime,DESC" // Return scans in descending order start with most recently completed.
+
+	finishedScanStatus = "finished" // Status for scans which have completed successfully.
+)
+
+type page struct {
+	Number         int `json:"number"`
+	Size           int `json:"size"`
+	TotalPages     int `json:"totalPages"`
+	TotalResources int `json:"totalResources"`
+}
+
+type resource struct {
+	EndTime string `json:"endTime"`
+	ScanID  int    `json:"id"`
+	SiteID  int    `json:"siteId"`
+	Status  string `json:"status"`
+}
+
+type nexposeScanResponse struct {
+	Page      page       `json:"page"`
+	Resources []resource `json:"resources"`
+}
+
+// NexposeClient implements the interfaces to fetch scans from Nexpose
+type NexposeClient struct {
+	Client   *http.Client
+	Host     *url.URL
+	PageSize int
+}
+
+// FetchScans fetches Nexpose scans, filters out running scans, and returns all completed scans
+// after the provided timestamp.
+func (n *NexposeClient) FetchScans(ctx context.Context, ts time.Time) ([]domain.CompletedScan, error) {
+	var completedScans []domain.CompletedScan
+
+	scanResp, err := n.makePagedNexposeScanRequest(0)
+	if err != nil {
+		return nil, err
+	}
+
+	pages := scanResp.Page.TotalPages
+	for _, resource := range scanResp.Resources {
+		completedScan, err := scanResourceToCompletedScan(resource, ts)
+		switch err.(type) {
+		case nil:
+			completedScans = append(completedScans, completedScan)
+		case outOfRangeError:
```

Comment:
Question: It's not immediately obvious to me here that we're ignoring errors from `scanResourceToCompletedScan` for unfinished scans/time parse errors. For the latter - that error looks like it won't actually be logged - how would we know that specific scan wasn't processed? Further, does that error signify that we should fail the current batch of scans? Do we have a way to trigger the processing of a specific scan?

Reply:
There's no way to currently process a specific scan. Looking over the Nexpose API docs, the timestamp should always be a valid ISO8601 timestamp, so we really should never hit the time parse error case. That said, I've added some additional comments and an error for the unfinished scan case, and added a default case for any other errors. That should ensure the handler logs any time parse errors.

Reply:
This looks good to me now, thanks!!